### ISSUE TITLE
Suggestion: Conversation Threading

### ISSUE THREAD

**Issue by mofosyne** [ircv3/ircv3-specifications#226]
Has there been any thoughts about supporting conversation threading? Is there any benefits to using it?

Basically, allows for referring to a previous message. So that an IRC conversation could be multithreaded like reddit.

There is some precedence of conversation threading used in real time instant messaging protocol, such as telegram. It seems to be a pretty useful feature. 

https://en.wikipedia.org/wiki/Conversation_threading

---

Possible references to this concept:
- https://github.com/flowdock/oulu/issues/47 "Replying to specific discussions is a useful feature of Flowdock, but there seems to be no way of doing it through IRC." -- gitkrax |  "@gitkrax Unfortunately the IRC protocol doesn't support threads, so we would need to do something that's based on plain text." -- mutru

**Comment by SaberUK** [comment]
Cool idea but I don't think its feasible to implement a sane interface for this considering IRC channel scrollback can be multiple megabytes in size.

**Comment by highflyer77** [comment]
I have very often thought of this and I think it would be a great idea!  
But I also don't know how it could be implemented in IRC.  Possibly the client could move a collapsed thread back down to the bottom of the message buffer somehow, maybe that could work.

**Comment by dequis** [comment]
Bitlbee's twitter module handles this by showing a two digit hex number before each tweet, for example, and marking replies with `[source->target]`, for example:

```
<foo> [aa] @bar hello
<bar> [ab->aa] @foo sup
<baz> [ac] other tweet
```

Only the last 256 tweets are remembered so older replies appear as if they aren't replying to anything.

On the protocol level it would be enough to have a tag for unique, server generated message IDs (because you can't trust client IDs to be unique enough), and another tag to specify what ID is being replied to.

**Comment by mofosyne** [comment]
to dequis : 
  Alternatively the ID could be a hash (doesn't have to be strong, even md5 could do).

-------

to SaberUK : 
  There are various ways to approach the User Interface aspect of real time threaded conversation.

* Jump to Prev/Next : In telegram, you can jump to previous comment, which is still held in cronological order. There is a button to then jump forward (and eventually back to real time view).

* Split screen: The mainscreen chatroom remains the same, however clicking on a thread would open up a split screen showing only the post in that thread (as if it was a separate chatroom). This is good for observing the main chatroom, while seeing any updates to the thread.

* Tree branch interface: Everything is arranged not in chronological order, but in a branching thread interface (like reddit). Good for overview, but not as good in keeping track of current chat.

Anyway the point is, there is various possible approaches to dealing with this, and it has been done in actual instant messaging apps already. (Heck 4chan can even count as one if you F4 refresh fast enough).

Either-way, the minimum needed to allow for these kinds of experimentation, is just to have the ability to tag a message as having a parent.

**Comment by DarthGandalf** [comment]
In any case, client needs to support messages which are not declared to be
response to some particular other message, because there will be clients
which don't support such threading, and they don't send any special tag.

2016-02-12 17:49 GMT+00:00 mofosyne <notifications@github.com>:

**Comment by mofosyne** [comment]
I imagine the default behaviour for non thread supporting IRC, would be similar to how IRC normally function these days, people just talking randomly. Which we somehow were able to live with all these years, thought with some confusion.

A small patch could be applied to alleviate this somewhat, by appending the username that you are responding to (which is how people psudo-thread, their conversation in IRC). If the server can detect that the client dont do threading, you could perhaps even make the server add this in automatically.

e.g.

```
user A : "And so the idea is this"
user B : "Hi all!"
user C : "@A: Oh right yea"
user A : "@B: hi"
user C : "@B: hi"
```

It's obviously not perfect, but its how 'threads' sorta occurs in IRC. Just gotta follow it mentally one by one.

**Comment by dequis** [comment]
>Alternatively the ID could be a hash of the message+username+timestamp (doesn't have to be strong, even md5 could do).

I'd rather leave that up to the implementation. It just needs to be unique-ish. It could be a reference to a globally sequential id in a server side log database, for example.

**Comment by Aerdan** [comment]
The mechanics of the id tag should be implementation-defined, yes, as long as the client doesn't see any duplication.

**Comment by DarthGandalf** [comment]
@mofosyne I was talking about the other direction: old client sends plain message to channel, new client receives it and needs to somehow sensibly display it in presence of all other threads.

**Comment by mofosyne** [comment]
DarthGandalf well then the 'parent id' remains null and is just displayed chronologically always (e.g. appended). Which is the same if you just post directly without trying to reply to a particular thread in a thread supported client.

However you could make an argument to always append any message that is sent as `<username> : <message>` to the last post sent by that username. Which I think is mostly the general convention for 'replying' to a person directly in public within a channel.

**Comment by jwheare** [comment]
This is addressed by #288
